### PR TITLE
#167377535 Email notification for User verification

### PR DESCRIPTION
**Pull request by Merkll**
**What does this PR do?**
This pull request sets up email notification which is then used to send a verification email to user on successful signup

**Description of Task to be completed**
- create an email sending module for application
- create a notification system using an event-driven architecture
- send verification email to the user on signup

**How should this be manually tested?**
- clone this branch
- create a user account
- a verification email will be sent to the registered account

**Any Background Context You want to provide?**
N/A

#### What are the relevant pivotal tracker stories?

[#167377535](https://pivotaltracker.com/story/show/167377535)

### CODE REVIEW

**Review by Jonathan4github — APPROVED** [review]
Nice Job

**Review by successgilli — APPROVED** [review]
Nice work.

**Review by Bastieno — APPROVED** [review]
Good work!

File: src/services/email.service.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { config } from 'dotenv';
+import sendgrid from '@sendgrid/mail';
+import Debug from 'debug';
+
+config();
+
+const debug = Debug('dev');
+
```
Comment (iverenshaguy) [inline src/services/email.service.js]:
```suggestion
```

File: src/services/email.service.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const { SENDGRID_API_KEY, SENDGRID_EMAIL, SENDGRID_ACTIVATION_TEMPLATE } = process.env;
+const templates = { accountActivation: SENDGRID_ACTIVATION_TEMPLATE };
+
+
+/**
+     * Initialises the event .
```
Comment (iverenshaguy) [inline src/services/email.service.js]:
```suggestion
/**
 * Initialises the event
```
Please ensure you align JSDoc items properly. Also always leave a new line between JSDoc properties e.g

```javascript
/**
 * Initialises the event

 * @function

 * @param {Object} message - The message payload.
 * @param {String} type - the type of message which corresponds to the mail template
 * @param {Object | Array} data - can be a list for bulk sending. must contain email parameter

 * @return {promise} - Returns a promise that get resolcved to a success object
 */
```

File: src/services/email.service.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     * @param {Object} message - The message payload.
+     * @param {String} type - the type of message which corresponds to the mail template
+     * @param {Object | Array} data - can be a list for bulk sending. must contain email parameter
```
Comment (iverenshaguy) [inline src/services/email.service.js]:
```suggestion
     * @param {Array} data - can be a list for bulk sending. must contain email parameter
```
Always send it as an array

File: src/services/email.service.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+const debug = Debug('dev');
+
+const { SENDGRID_API_KEY, SENDGRID_EMAIL, SENDGRID_ACTIVATION_TEMPLATE } = process.env;
+const templates = { accountActivation: SENDGRID_ACTIVATION_TEMPLATE };
+
+
+/**
+     * Initialises the event .
+     * @param {Object} message - The message payload.
```
Comment (iverenshaguy) [inline src/services/email.service.js]: This wasn't used in the function or am I missing something?

File: src/services/email.service.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     * @return {promise} - Returns a promise that get resolcved to a success object
+*/
+export default async ({ type, data }) => {
+  const template = templates[type];
+
+  if (!template) throw new Error('Email Template not available');
+  if (!data) throw new Error('Mail data cannot be empty');
+  const mailDetails = Array.isArray(data) ? data : [data];
+
+  sendgrid.setApiKey(SENDGRID_API_KEY);
```
Comment (iverenshaguy) [inline src/services/email.service.js]: Kindly move this to the top of the function

File: src/services/email.service.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { config } from 'dotenv';
```
Comment (iverenshaguy) [inline src/services/email.service.js]: Can't remember us agreeing to use this format for our file names `email.service`. If you are going to use this, then you'll have to change all non-generated files to the same format. Our codebase has to be consistent

File: src/services/email.service.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+     * Initialises the event .
+     * @param {Object} message - The message payload.
+     * @param {String} type - the type of message which corresponds to the mail template
+     * @param {Object | Array} data - can be a list for bulk sending. must contain email parameter
+     * @return {promise} - Returns a promise that get resolcved to a success object
+*/
+export default async ({ type, data }) => {
+  const template = templates[type];
+
+  if (!template) throw new Error('Email Template not available');
```
Comment (iverenshaguy) [inline src/services/email.service.js]: Is there a more elegant way to handle this error? Or is something else catching it?

File: src/services/email.service.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  if (!data) throw new Error('Mail data cannot be empty');
+  const mailDetails = Array.isArray(data) ? data : [data];
+
+  sendgrid.setApiKey(SENDGRID_API_KEY);
+  const messagePayload = mailDetails.map((singleEmailData) => {
+    const { email } = singleEmailData;
+    return {
```
Comment (iverenshaguy) [inline src/services/email.service.js]: Please leave a new line before return statements

File: src/services/email.service.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      to: email,
+      from: SENDGRID_EMAIL,
+      templateId: template,
+      dynamic_template_data: {
+        ...singleEmailData
+      }
+    };
+  });
+  return sendgrid.send(messagePayload)
```
Comment (iverenshaguy) [inline src/services/email.service.js]: Same a above

File: src/services/email.service.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  const mailDetails = Array.isArray(data) ? data : [data];
+
+  sendgrid.setApiKey(SENDGRID_API_KEY);
+  const messagePayload = mailDetails.map((singleEmailData) => {
+    const { email } = singleEmailData;
+    return {
+      to: email,
+      from: SENDGRID_EMAIL,
+      templateId: template,
+      dynamic_template_data: {
```
Comment (iverenshaguy) [inline src/services/email.service.js]: I don't understand. Does the Sendgrid API require camel case and snake case in the same payload?

File: src/services/notification.service.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
+import EventEmitter from 'events';
+import emailService from './email.service';
+
+/**
+ * Class used for all email related notifications.
+ * @extends EventEmitter
```
Comment (iverenshaguy) [inline src/services/notification.service.js]: Good job. You can also add the @class keyword.

File: src/services/notification.service.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    super();
+    this.on('notification', this.handleNotification);
+  }
+
+  /**
+     * Initialises the event .
+     * @param {object} eventPayload - The event payload. Contains notification type and payload.
+     * @return {void}
+     */
+  handleNotification(eventPayload) {
```
Comment (iverenshaguy) [inline src/services/notification.service.js]:
```suggestion
  handleNotification({ type, payload }) {
```

File: src/services/notification.service.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+     * Handles notification for account activation.
+     * @param {String} type - type of mail being sent.
+     * @param {Object} data - message data.
+     * @return {object} - status of event executed
+     */
+  sendMail(type, data) {
+    emailService({ type, data });
+    const response = { status: 'sent' };
+    this.emit('notificationSent', { type, data });
+    return response;
```
Comment (iverenshaguy) [inline src/services/notification.service.js]:
```suggestion

    return { status: 'sent' };
```

File: src/services/notification.service.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+     * Handles notification for account activation.
+     * @param {String} type - type of mail being sent.
+     * @param {Object} data - message data.
+     * @return {object} - status of event executed
+     */
+  sendMail(type, data) {
+    emailService({ type, data });
+    const response = { status: 'sent' };
```
Comment (iverenshaguy) [inline src/services/notification.service.js]:
```suggestion
```

File: src/services/notification.service.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    this.on('notification', this.handleNotification);
+  }
+
+  /**
+     * Initialises the event .
+     * @param {object} eventPayload - The event payload. Contains notification type and payload.
+     * @return {void}
+     */
+  handleNotification(eventPayload) {
+    const { type, payload } = eventPayload;
```
Comment (iverenshaguy) [inline src/services/notification.service.js]:
```suggestion
```

File: src/services/notification.service.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+     */
+  sendMail(type, data) {
+    emailService({ type, data });
+    const response = { status: 'sent' };
+    this.emit('notificationSent', { type, data });
+    return response;
+  }
+}
+
+
```
Comment (iverenshaguy) [inline src/services/notification.service.js]:
```suggestion
```

File: tests/services/email.test.js [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+import sinonTest from 'sinon-test';
+import sendgrid from '@sendgrid/mail';
+
+import Email from '../../src/services/email.service';
+
+chai.use(chaiAsPromised);
+
+const test = sinonTest(sinon);
+
+chai.should();
```
Comment (iverenshaguy) [inline tests/services/email.test.js]: Kindly move to line 12 and leave a new line below it

File: tests/services/email.test.js [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+  it('Should throw an error if template is not available', async () => {
+    (Email({ type: 'unAvailable', data: emailData })).should.eventually.be.rejected;
+  });
+
+  it('Should send message with single mail data object', test(async () => {
+    sinon.stub(sendgrid, 'send').resolves();
+
+    const response = await Email({ type: 'accountActivation', data: emailData });
+    sendgrid.send.restore();
```
Comment (iverenshaguy) [inline tests/services/email.test.js]: Kindly restore in an `afterEach` hook.

File: tests/services/email.test.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
+/* eslint-disable no-unused-expressions */
+import chai from 'chai';
```
Comment (iverenshaguy) [inline tests/services/email.test.js]: Can we do 
```suggestion
import { should } from 'chai';
```
and then `should()`?

File: tests/services/notification.test.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
+import { should as importedShould } from 'chai';
+
+import notification from '../../src/services/notification.service';
+
+const should = importedShould();
+
+describe('Notification System', () => {
+  it('Should return undefined for unsupported types', () => {
```
Comment (iverenshaguy) [inline tests/services/notification.test.js]:
```suggestion
  it('should return undefined for unsupported types', () => {
```
Kindly correct in all your files

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
Good job. Kindly review my comments

File: src/services/email.service.js (hunk 2 above)
Comment (Merkll) [reply src/services/email.service.js]: eslint enforces the indentation I used

**Comment by Merkll** [comment]
I have reviewed your comments

File: src/services/email.service.js (hunk 2 above)
Comment (iverenshaguy) [reply src/services/email.service.js]: Can I see a screenshot?

File: src/services/email.service.js (hunk 7 above)
Comment (Merkll) [reply src/services/email.service.js]: it is caught when the function is consumed

File: src/services/email.service.js (hunk 10 above)
Comment (Merkll) [reply src/services/email.service.js]: yes the dynamic template data is snakecase

File: src/services/email.service.js (hunk 7 above)
Comment (Merkll) [reply src/services/email.service.js]: I think will be better to return the error rather than throwing it
Or what do you think?

**Review by Merkll — COMMENTED** [review]
implemented

File: src/services/email.service.js (hunk 2 above)
Comment (Merkll) [reply src/services/email.service.js]: Screenshot 2019-07-30 at 12 23 48 PM

Screenshot of Jsdoc indentation and Eslint red flag

File: tests/services/email.test.js (hunk 19 above)
Comment (Merkll) [reply tests/services/email.test.js]: i needed the chai to use chai-as-promised

**Comment by Merkll** [comment]
I have implemented the feedback

**Review by iverenshaguy — APPROVED** [review]
Great work, thank you.

**Pull request merged**